### PR TITLE
Change escape character from '$' to '\' for `regexp_replace`

### LINKED ISSUE #7664: Need to change escape character from '$' to '\' for `regexp_replace`

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #7664

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Common/OptimizedRegularExpression.inl.h [hunk 1]
```diff
@@ -517,8 +511,8 @@ void OptimizedRegularExpressionImpl<thread_safe>::processReplaceEmptyStringExpr(
     }
 
     StringPieceType expr_sp(subject, subject_size);
-    StringPieceType matches[MAX_CAPTURES];
-    bool success = re2->Match(expr_sp, 0, expr_sp.size(), re2_st::RE2::Anchor::UNANCHORED, matches, capture_num);
+    StringPieceType matches[capture_num+1];
```
Comment (gengliqi) [inline dbms/src/Common/OptimizedRegularExpression.inl.h:514]: Can this code compile?

Comment (xzhangxian1008) [reply dbms/src/Common/OptimizedRegularExpression.inl.h]: > Does this code compile?

I have deleted these codes.

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**